### PR TITLE
Add `dynamic_one_shot` tensorflow support + expanded testing

### LINKED ISSUE #5736: [BUG] Differentiation does not work with `dynamic_one_shot`
### Expected behavior

I expect to be able to differentiate arbitrary circuits when using the `dynamic_one_shot` transform.

### Actual behavior

I get an error.

### Additional information

Originally from this [forum discussion](https://discuss.pennylane.ai/t/running-circuits-with-many-mid-circuit-measurements/4355/15)

### Source code

```shell
import pennylane as qml
import torch

dev = qml.device("default.qubit", shots=10)

@qml.qnode(dev, interface='torch') # switch to torch interface
def f(x):
    qml.RX(x, 0) # remove extraneous instructions
    return qml.expval(qml.measure(0)) # REPLACE PauliX with measure

x = torch.tensor(0.4, requires_grad=True) # switch to torch tensor
result = f(x) 
result.backward() # replace with torch gradient computation
x.grad
```

### Tracebacks

```shell
/usr/local/lib/python3.10/dist-packages/autoray/autoray.py:81: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  return func(*args, **kwargs)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-879285d05025> in <cell line: 13>()
     11 x = torch.tensor(0.4, requires_grad=True) # switch to torch tensor
     12 result = f(x)
---> 13 result.backward() # replace with torch gradient computation
     14 x.grad

1 frames
/usr/local/lib/python3.10/dist-packages/torch/_tensor.py in backward(self, gradient, retain_graph, create_graph, inputs)
    520                 inputs=inputs,
    521             )
--> 522         torch.autograd.backward(
    523             self, gradient, retain_graph, create_graph, inputs=inputs
    524         )

/usr/local/lib/python3.10/dist-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)
    264     # some Python versions print out the first line of a multi-line function
    265     # calls in the traceback and some print out the last line
--> 266     Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
    267         tensors,
    268         grad_tensors_,

RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

### System information

```shell
Dev. Using Pennylane branch dos-interfaces
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
As name says. Gradient workflows no longer raise errors after the merge of #5791 , but their correctness is yet to be verified.

**Description of the Change:**
* Add differentiation tests for jax, torch, autograd, tensorflow

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
Fixes #5736, #5710

### CODE REVIEW

**Comment by mudit2812** [comment]
[sc-63943] [sc-63618]

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 1]
```diff
@@ -424,156 +424,224 @@ def circuit(x):
...
-    results1 = qml.QNode(func, dev, mcm_method="deferred")(*params)
-
-    mcm_utils.validate_measurements(qml.sample, shots, results1, results0, batch_size=None)
+class TestJaxIntegration:
+    """Integration tests for dynamic_one_shot with jax"""
+
+    @pytest.mark.parametrize("mcm_method", ["one-shot", "tree-traversal"])
+    @pytest.mark.parametrize("shots", [100, [100, 101], [100, 100, 101]])
+    @pytest.mark.parametrize("postselect", [None, 0, 1])
+    def test_sample_with_prng_key(self, mcm_method, shots, postselect):
```
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:435]: This test and the one below it have just been moved into the `TestJaxIntegration` class so no need to be thorough reviewing the aforementioned two tests

File: pennylane/transforms/dynamic_one_shot.py [hunk 2]
```diff
@@ -248,18 +248,25 @@ def measurement_with_no_shots(measurement):
...
-    mcm_samples = qml.math.hstack(tuple(res.reshape((-1, 1)) for res in results[-n_mcms:]))
+    mcm_samples = qml.math.hstack(
+        tuple(qml.math.reshape(res, (-1, 1)) for res in results[-n_mcms:])
+    )
     mcm_samples = qml.math.array(mcm_samples, like=interface)
     # Can't use boolean dtype array with tf, hence why conditionally setting items to 0 or 1
     has_postselect = qml.math.array(
-        [[int(op.postselect is not None) for op in all_mcms]], like=interface
+        [[int(op.postselect is not None) for op in all_mcms]],
+        like=interface,
```
Comment (albi3ro) [inline pennylane/transforms/dynamic_one_shot.py:263]: I don't think we want to convert back to interface data here.  If we call `dynamic_one_shot` after we've converted to numpy, we don't want to convert back into interface data.

Right now it looks like we are still calling `convert_to_numpy_parameters` after device preprocessing, but I could very easily see that changing in the future.

I was under the assumption that `interface` was passed to this function for the purpose of determining  jit-specific postselection handling.

Comment (albi3ro) [reply pennylane/transforms/dynamic_one_shot.py:263]: Maybe we could have a "postselection interface" and "result data interface"?

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 3]
```diff
@@ -424,156 +425,141 @@ def circuit(x):
...
-def test_sample_with_prng_key(mcm_method, shots, postselect):
-    """Test that setting a PRNGKey gives the expected behaviour. With separate calls
-    to DefaultQubit.execute, the same results are expected when using a PRNGKey"""
-    # pylint: disable=import-outside-toplevel
-    from jax.random import PRNGKey
-
-    dev = get_device(shots=shots, seed=PRNGKey(678))
-    params = [np.pi / 4, np.pi / 3]
-    obs = qml.PauliZ(0) @ qml.PauliZ(1)
+def test_finite_diff_in_transform_program(interface, mcm_method):
```
Comment (albi3ro) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:431]: I'm not entirely sure this is a mcm test if the diff method is explicitly specified as `"finite-diff"`.  What happens if `diff_method="best"`? Should we automatically switch to finite diff, or give a warning?

Comment (mudit2812) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:431]: Then we use parameter shift, but parameter shift is not correct if there is postselection or classical control.

Comment (mudit2812) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:431]: > Should we automatically switch to finite diff, or give a warning?

I think it makes sense to switch to finite diff.

File: pennylane/transforms/dynamic_one_shot.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/transforms/dynamic_one_shot.py:263]: > I was under the assumption that interface was passed to this function for the purpose of determining jit-specific postselection handling.

Yes, for pennylane that is the case, so I can definitely spend time investigating if we really need to convert to interface data, but for Catalyst we need to propagate jax data correctly.

**Pull request closed without merging**